### ISSUE TITLE
Dart 2.16 seg fault whilst debugging

### ISSUE THREAD

**Issue by bsutton** [dart-lang/sdk#48024]
Seg fault whilst debugging.

A couple of notes for clarity.

These are the dcli unit tests. The spawn dart pub get hence the 'Resolving dependencies...' message.
I don't think this relates to the issue.
You can ignore 'dswitch' in any paths, this is just the tool I use to switch between versions of dart.

dart --version
Dart SDK version: 2.16.0-80.1.beta (beta) (Mon Dec 13 11:59:02 2021 +0100) on "linux_x64"

This occured whilst stepping through a debug session in vscode. I had run the same debug session a number of times before this occured.
Haven't been able to reproduce it since.

```
Resolving dependencies...
Because dcli_unit_tester depends on dcli_core from path which doesn't exist (could not find package dcli_core at "../../../../dcli_core"), version solving failed.
Connecting to VM Service at http://127.0.0.1:33473/q2VuVzjkpfQ=/ws
[38;5;208;1mRe-using common TestFileSystem
[31;1m++++++++++++++++++++Creating TestFileSystem /tmp/dcli/90c3b785-7d4c-4e69-a412-1eba3c9e425e for isolate isolates/2187070800850199++++++++++++++++++++
[31;1m**************************************** Starting test File: '///home/bsutton/git/dcli/dcli/test/src/apps/dcli_complete_test.dart':84 ********************************************************************************
[32;1mUsing TestFileSystem /tmp/dcli/90c3b785-7d4c-4e69-a412-1eba3c9e425e for Isolate: isolates/2187070800850199
Reset dcliPath: /tmp/dcli/90c3b785-7d4c-4e69-a412-1eba3c9e425e/.dcli

===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=0xa7
version=2.16.0-80.1.beta (beta) (Mon Dec 13 11:59:02 2021 +0100) on "linux_x64"
pid=3132350, thread=3132393, isolate_group=main(0x56320214e800), isolate=main(0x56320224d000)
isolate_instructions=563200ed32e0, vm_instructions=563200ed32e0
  pc 0x00005632010cf01c fp 0x00007f7295bfbb90 dart::Class::EnsureDeclarationLoaded() const+0xc
  pc 0x000056320103ffdd fp 0x00007f7295bfbc00 dart::ClassFinalizer::ExpandAndFinalizeTypeArguments(dart::Zone*, dart::AbstractType const&, dart::ZoneGrowableHandlePtrArray<dart::AbstractType const>*)+0x3d
  pc 0x0000563201040580 fp 0x00007f7295bfbc50 dart::ClassFinalizer::FinalizeType(dart::AbstractType const&, dart::ClassFinalizer::FinalizationKind, dart::ZoneGrowableHandlePtrArray<dart::AbstractType const>*)+0x320
  pc 0x00005632010c48fa fp 0x00007f7295bfbce0 dart::Class::DeclarationType() const+0x2fa
  pc 0x0000563201564e9e fp 0x00007f7295bfbd20 dart::kernel::TranslationHelper::GetDeclarationType(dart::Class const&)+0xde
  pc 0x0000563201574079 fp 0x00007f7295bfbea0 dart::kernel::ScopeBuilder::BuildScopes()+0x549
  pc 0x00005632011347d7 fp 0x00007f7295bfc1e0 dart::ParsedFunction::EnsureKernelScopes()+0x37
  pc 0x000056320153eff3 fp 0x00007f7295bfc210 dart::kernel::StreamingFlowGraphBuilder::ParseKernelASTFunction()+0x53
  pc 0x000056320153ee51 fp 0x00007f7295bfc2c0 dart::kernel::StreamingFlowGraphBuilder::BuildGraph()+0xb1
  pc 0x0000563201552043 fp 0x00007f7295bfc570 dart::kernel::FlowGraphBuilder::BuildGraph()+0x63
  pc 0x00005632011c9f90 fp 0x00007f7295bfc7b0 dart::DartCompilationPipeline::BuildFlowGraph(dart::Zone*, dart::ParsedFunction*, dart::ZoneGrowableArray<dart::ICData const*>*, long, bool)+0x40
  pc 0x00005632011cb02f fp 0x00007f7295bfceb0 dart::CompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0x47f
  pc 0x00005632011cbf79 fp 0x00007f7295bfd0c0 /home/bsutton/.dswitch/channels/beta/versions/2.16.0-80.1.beta/dart-sdk/bin/dart+0x20eaf79
  pc 0x00005632011cba6a fp 0x00007f7295bfd150 dart::Compiler::CompileFunction(dart::Thread*, dart::Function const&)+0x22a
  pc 0x000056320104eea7 fp 0x00007f7295bfd1b0 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)+0xb7
  pc 0x00005632010cece8 fp 0x00007f7295bfd5a0 /home/bsutton/.dswitch/channels/beta/versions/2.16.0-80.1.beta/dart-sdk/bin/dart+0x1fedce8
  pc 0x000056320119b7ce fp 0x00007f7295bfd620 /home/bsutton/.dswitch/channels/beta/versions/2.16.0-80.1.beta/dart-sdk/bin/dart+0x20ba7ce
  pc 0x00005632011922a9 fp 0x00007f7295bfd760 dart::Service::InvokeMethod(dart::Isolate*, dart::Array const&, bool)+0x329
  pc 0x00005632011929a1 fp 0x00007f7295bfd780 dart::Service::HandleIsolateMessage(dart::Isolate*, dart::Array const&)+0x11
  pc 0x0000563201077f50 fp 0x00007f7295bfd860 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x280
  pc 0x000056320109ff5c fp 0x00007f7295bfd8d0 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x14c
  pc 0x00005632010a0387 fp 0x00007f7295bfd910 dart::MessageHandler::HandleOOBMessages()+0x47
  pc 0x00005632011b6258 fp 0x00007f7295bfd940 dart::Thread::HandleInterrupts()+0x108
  pc 0x0000563201189d82 fp 0x00007f7295bfdaf0 dart::DRT_StackOverflow(dart::NativeArguments)+0x812
  pc 0x00007f729e182573 fp 0x00007f7295bfdb30 Unknown symbol
  pc 0x00007f729bc8047f fp 0x00007f7295bfdb68 Unknown symbol
  pc 0x00007f729bca3afb fp 0x00007f7295bfdba8 Unknown symbol
  pc 0x00007f729bca37e5 fp 0x00007f7295bfdbe0 Unknown symbol

  pc 0x00007f729bc93b8d fp 0x00007f7295bfdc68 Unknown symbol
  pc 0x00007f729bc9317e fp 0x00007f7295bfdcb8 Unknown symbol
  pc 0x00007f729517884c fp 0x00007f7295bfdcf0 Unknown symbol
  pc 0x00007f7295176760 fp 0x00007f7295bfdd88 Unknown symbol
  pc 0x00007f7297b52079 fp 0x00007f7295bfde00 Unknown symbol
  pc 0x00007f7297b51f0c fp 0x00007f7295bfde68 Unknown symbol
  pc 0x00007f7297b51c66 fp 0x00007f7295bfdf00 Unknown symbol
  pc 0x00007f7297b59d2e fp 0x00007f7295bfdf50 Unknown symbol
  pc 0x00007f7297b59a4e fp 0x00007f7295bfdfb0 Unknown symbol
  pc 0x00007f7297b58d57 fp 0x00007f7295bfe030 Unknown symbol
  pc 0x00007f7297b57a7e fp 0x00007f7295bfe078 Unknown symbol
  pc 0x00007f729515ee10 fp 0x00007f7295bfe0b8 Unknown symbol
  pc 0x00007f72951772d3 fp 0x00007f7295bfe158 Unknown symbol
  pc 0x00007f7297b52079 fp 0x00007f7295bfe1d0 Unknown symbol
  pc 0x00007f7297b51f0c fp 0x00007f7295bfe238 Unknown symbol
  pc 0x00007f7297b51c66 fp 0x00007f7295bfe2d0 Unknown symbol
  pc 0x00007f7297b59d2e fp 0x00007f7295bfe320 Unknown symbol
  pc 0x00007f7297b59a4e fp 0x00007f7295bfe380 Unknown symbol
  pc 0x00007f7297b58d57 fp 0x00007f7295bfe400 Unknown symbol
  pc 0x00007f729510e038 fp 0x00007f7295bfe440 Unknown symbol
  pc 0x00007f7297b41d71 fp 0x00007f7295bfe4b0 Unknown symbol
  pc 0x00007f7297b41ad9 fp 0x00007f7295bfe510 Unknown symbol
  pc 0x00007f7297b413e3 fp 0x00007f7295bfe5a0 Unknown symbol
  pc 0x00007f7297b576ef fp 0x00007f7295bfe610 Unknown symbol
  pc 0x00007f7297b575ed fp 0x00007f7295bfe650 Unknown symbol
  pc 0x00007f7297b57373 fp 0x00007f7295bfe698 Unknown symbol
  pc 0x00007f7297b56feb fp 0x00007f7295bfe6d8 Unknown symbol
  pc 0x00007f7297b56f19 fp 0x00007f7295bfe700 Unknown symbol
  pc 0x00007f7297b4fe8e fp 0x00007f7295bfe740 Unknown symbol
  pc 0x00007f729516f357 fp 0x00007f7295bfe7c0 Unknown symbol
  pc 0x00007f729516d6c8 fp 0x00007f7295bfe7f8 Unknown symbol
  pc 0x00007f729516d3dc fp 0x00007f7295bfe828 Unknown symbol
  pc 0x00007f7297b1f439 fp 0x00007f7295bfe890 Unknown symbol
  pc 0x00007f7297b1eca3 fp 0x00007f7295bfe8e0 Unknown symbol
  pc 0x00007f729e1829bf fp 0x00007f7295bfe958 Unknown symbol
  pc 0x000056320104f0fd fp 0x00007f7295bfea00 dart::DartEntry::InvokeCode(dart::Code const&, unsigned long, dart::Array const&, dart::Array const&, dart::Thread*)+0x14d
  pc 0x000056320104ef3c fp 0x00007f7295bfea60 dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)+0x14c
  pc 0x000056320105149f fp 0x00007f7295bfeab0 dart::DartLibraryCalls::HandleMessage(long, dart::Instance const&)+0x16f
  pc 0x0000563201078045 fp 0x00007f7295bfeb90 dart::IsolateMessageHandler::HandleMessage(std::__2::unique_ptr<dart::Message, std::__2::default_delete<dart::Message> >)+0x375
  pc 0x000056320109ff5c fp 0x00007f7295bfec00 dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)+0x14c
  pc 0x00005632010a065f fp 0x00007f7295bfec50 dart::MessageHandler::TaskCallback()+0x1df
  pc 0x00005632011b9b58 fp 0x00007f7295bfece0 dart::ThreadPool::WorkerLoop(dart::ThreadPool::Worker*)+0x138
  pc 0x00005632011b9fec fp 0x00007f7295bfed10 dart::ThreadPool::Worker::Main(unsigned long)+0x5c
  pc 0x0000563201133908 fp 0x00007f7295bfedd0 /home/bsutton/.dswitch/channels/beta/versions/2.16.0-80.1.beta/dart-sdk/bin/dart+0x2052908
-- End of DumpStackTrace
  pc 0x0000000000000000 fp 0x00007f7295bfdb30 sp 0x0000000000000000 [Stub] CallToRuntime
  pc 0x00007f729bc8047f fp 0x00007f7295bfdb68 sp 0x00007f7295bfdb40 [Unoptimized] print
  pc 0x00007f729bca3afb fp 0x00007f7295bfdba8 sp 0x00007f7295bfdb78 [Unoptimized] TestFileSystem.copyTestScripts
  pc 0x00007f729bca37e5 fp 0x00007f7295bfdbe0 sp 0x00007f7295bfdbb8 [Unoptimized] TestFileSystem.initFS
  pc 0x00007f729bc93b8d fp 0x00007f7295bfdc68 sp 0x00007f7295bfdbf0 [Unoptimized] TestFileSystem._runUnderLock@21122130
  pc 0x00007f729bc9317e fp 0x00007f7295bfdcb8 sp 0x00007f7295bfdc78 [Unoptimized] TestFileSystem.withinZone
  pc 0x00007f729517884c fp 0x00007f7295bfdcf0 sp 0x00007f7295bfdcc8 [Unoptimized] main.<anonymous closure>.<anonymous closure>
  pc 0x00007f7295176760 fp 0x00007f7295bfdd88 sp 0x00007f7295bfdd00 [Unoptimized] Declarer.test.<anonymous closure>.<anonymous closure>
  pc 0x00007f7297b52079 fp 0x00007f7295bfde00 sp 0x00007f7295bfdd98 [Unoptimized] _rootRunUnary@4048458
  pc 0x00007f7297b51f0c fp 0x00007f7295bfde68 sp 0x00007f7295bfde10 [Unoptimized] _rootRunUnary@4048458
  pc 0x00007f7297b51c66 fp 0x00007f7295bfdf00 sp 0x00007f7295bfde78 [Unoptimized] _CustomZone@4048458.runUnary
  pc 0x00007f7297b59d2e fp 0x00007f7295bfdf50 sp 0x00007f7295bfdf10 [Unoptimized] _FutureListener@4048458.handleValue
  pc 0x00007f7297b59a4e fp 0x00007f7295bfdfb0 sp 0x00007f7295bfdf60 [Unoptimized] _Future@4048458._propagateToListeners@4048458.handleValueCallback
  pc 0x00007f7297b58d57 fp 0x00007f7295bfe030 sp 0x00007f7295bfdfc0 [Unoptimized] _Future@4048458._propagateToListeners@4048458
  pc 0x00007f7297b57a7e fp 0x00007f7295bfe078 sp 0x00007f7295bfe040 [Unoptimized] _Future@4048458._completeWithValue@4048458
  pc 0x00007f729515ee10 fp 0x00007f7295bfe0b8 sp 0x00007f7295bfe088 [Unoptimized] _completeOnAsyncReturn@4048458
  pc 0x00007f72951772d3 fp 0x00007f7295bfe158 sp 0x00007f7295bfe0c8 [Unoptimized] Declarer._runSetUps@412455614
  pc 0x00007f7297b52079 fp 0x00007f7295bfe1d0 sp 0x00007f7295bfe168 [Unoptimized] _rootRunUnary@4048458
  pc 0x00007f7297b51f0c fp 0x00007f7295bfe238 sp 0x00007f7295bfe1e0 [Unoptimized] _rootRunUnary@4048458
  pc 0x00007f7297b51c66 fp 0x00007f7295bfe2d0 sp 0x00007f7295bfe248 [Unoptimized] _CustomZone@4048458.runUnary
  pc 0x00007f7297b59d2e fp 0x00007f7295bfe320 sp 0x00007f7295bfe2e0 [Unoptimized] _FutureListener@4048458.handleValue
  pc 0x00007f7297b59a4e fp 0x00007f7295bfe380 sp 0x00007f7295bfe330 [Unoptimized] _Future@4048458._propagateToListeners@4048458.handleValueCallback
  pc 0x00007f7297b58d57 fp 0x00007f7295bfe400 sp 0x00007f7295bfe390 [Unoptimized] _Future@4048458._propagateToListeners@4048458
  pc 0x00007f729510e038 fp 0x00007f7295bfe440 sp 0x00007f7295bfe410 [Unoptimized] _Future@4048458._addListener@4048458.<anonymous closure>
  pc 0x00007f7297b41d71 fp 0x00007f7295bfe4b0 sp 0x00007f7295bfe450 [Unoptimized] _rootRun@4048458
  pc 0x00007f7297b41ad9 fp 0x00007f7295bfe510 sp 0x00007f7295bfe4c0 [Unoptimized] _rootRun@4048458
  pc 0x00007f7297b413e3 fp 0x00007f7295bfe5a0 sp 0x00007f7295bfe520 [Unoptimized] _CustomZone@4048458.run
  pc 0x00007f7297b576ef fp 0x00007f7295bfe610 sp 0x00007f7295bfe5b0 [Unoptimized] _CustomZone@4048458.runGuarded
  pc 0x00007f7297b575ed fp 0x00007f7295bfe650 sp 0x00007f7295bfe620 [Unoptimized] _CustomZone@4048458.bindCallbackGuarded.<anonymous closure>
  pc 0x00007f7297b57373 fp 0x00007f7295bfe698 sp 0x00007f7295bfe660 [Unoptimized] _microtaskLoop@4048458
  pc 0x00007f7297b56feb fp 0x00007f7295bfe6d8 sp 0x00007f7295bfe6a8 [Unoptimized] _startMicrotaskLoop@4048458
  pc 0x00007f7297b56f19 fp 0x00007f7295bfe700 sp 0x00007f7295bfe6e8 [Unoptimized] _startMicrotaskLoop@4048458
  pc 0x00007f7297b4fe8e fp 0x00007f7295bfe740 sp 0x00007f7295bfe710 [Unoptimized] _runPendingImmediateCallback@1026248
  pc 0x00007f729516f357 fp 0x00007f7295bfe7c0 sp 0x00007f7295bfe750 [Unoptimized] _Timer@1026248._runTimers@1026248
  pc 0x00007f729516d6c8 fp 0x00007f7295bfe7f8 sp 0x00007f7295bfe7d0 [Unoptimized] _Timer@1026248._handleMessage@1026248
  pc 0x00007f729516d3dc fp 0x00007f7295bfe828 sp 0x00007f7295bfe808 [Unoptimized] _Timer@1026248._handleMessage@1026248
  pc 0x00007f7297b1f439 fp 0x00007f7295bfe890 sp 0x00007f7295bfe838 [Unoptimized] _Closure@0150898.dyn:call
  pc 0x00007f7297b1eca3 fp 0x00007f7295bfe8e0 sp 0x00007f7295bfe8a0 [Unoptimized] _RawReceivePortImpl@1026248._handleMessage@1026248
  pc 0x00007f729e1829bf fp 0x00007f7295bfe958 sp 0x00007f7295bfe8f0 [Stub] InvokeDartCode
=== Crash occured when compiling dart:core_::__Eval in unoptimized JIT mode in unknown pass
=== Flow Graph not available

Aborted (core dumped)
Exited (134)

```

**Comment by alexmarkov** [comment]
Here is what I was able to figure out from the stack trace and peeking into the code:

Crash happens while compiling a function for the expression evaluation (`:Eval`). Expression evaluation function is placed into a synthetic class (`#DebugClass`) which is not properly registered in the VM and has `kIllegalCid` class id (to make sure that class can be garbage collected).

ScopeBuilder calls `TranslationHelper::GetDeclarationType` because expression evaluation is performed in the context of an instance method.

For some reason `#DebugClass` was type finalized by the time ScopeBuilder is called during compilation of `:Eval` function. As a result, `TranslationHelper::GetDeclarationType` calls `Class::DeclarationType`, which creates an instance of `Type` corresponding to `#DebugClass`, and stores its class id (kIllegalCid) in the `Type`. After that, `ClassFinalizer::FinalizeType` is called on this `Type` which calls `ClassFinalizer::ExpandAndFinalizeTypeArguments` where class is obtained from the `Type` using `Type::type_class()` which doesn't work for kIllegalCid and causes crash.

It is not clear what caused finalization of `#DebugClass`. In the examples I tried the expression evaluation class `#DebugClass` is always not finalized, so I wasn't able to reproduce the problem so far. However, we can probably add some safeguards against expression evaluation classes to `TranslationHelper::GetDeclarationType` to make sure `Class::DeclarationType` is not called for the classes which are not properly registered.
